diff --git a/arch/mips/pci/pci-legacy.c b/arch/mips/pci/pci-legacy.c
index 39052de91..9fc9e9409 100644
--- a/arch/mips/pci/pci-legacy.c
+++ b/arch/mips/pci/pci-legacy.c
@@ -82,7 +82,7 @@ static void pcibios_scanbus(struct pci_controller *hose)
 	if (!bridge)
 		return;
 
-	if (hose->get_busno && pci_has_flag(PCI_PROBE_ONLY))
+	if (hose->get_busno)
 		next_busno = (*hose->get_busno)();
 
 	pci_add_resource_offset(&resources,
